_posts/current/blog/2022-01-06-how-to-emulate-trusted-platform-module-in-qemu-with-u-boot.md

```diff
@@ -0,0 +1,101 @@
+
+On typical measured boot scenarios the firmware (or even BL1 is some devices) hashes itself to specific PCRs. So you can have for example BL2 measure itself and then measure BL31, BL32, BL33 and even GRUB config files,  commands, Linux initrd etc.  This effectively creates a chain of trust, which guarantees system components that we choose to measure have not been modified. Changing any of these would result in a different set of hash values.
+
+## U-Boot Support
+
+There is a specification defining a [standard](https://trustedcomputinggroup.org/wp-content/uploads/EFI-Protocol-Specification-rev13-160330final.pdf) interface to the TPM on an UEFI platform. Its purpose is to define APIs and provide information, for things like, is a TPM present, which PCR banks are active, change active PCR banks, obtain the TCG boot log, extend hashes to PCRs,  append events to the TCG boot log etc.
+U-Boot recently got support for this, as well as support for the [TCG PC Client Platform Firmware Profile](https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClient_PFP_r1p05_v22_02dec2020.pdf).
+Patches for U-Boot were contributed by Linaro and can be found [here](https://lore.kernel.org/u-boot/20201112222210.876652-1-ilias.apalodimas@linaro.org/), [here](https://lore.kernel.org/u-boot/20201127162932.1965323-1-ilias.apalodimas@linaro.org/) and [here](https://lore.kernel.org/u-boot/20210813071243.18885-1-masahisa.kojima@linaro.org/).
+It can be tricky to find an Arm device with a TPMv2. If you have a board with an RPMB and OP-TEE support, we recommend trying Microsoft's [fTPM](https://github.com/microsoft/ms-tpm-20-ref/). However testing that in QEMU won't work since it lacks RPMB emulation. Luckily there is another solution.
+
+## Using SWTPM
+
+[SWTPM](https://github.com/stefanberger/swtpm) is a TPM emulator that works under QEMU. It provides a memory mapped device which adheres to the [TCG TPM Interface Specification](https://trustedcomputinggroup.org/wp-content/uploads/TCG_PCClientTPMInterfaceSpecification_TIS__1-3_27_03212013.pdf).  U-Boot lacked an MMIO TPMv2 driver up until [this patchset](https://source.denx.de/u-boot/u-boot/-/commit/e0ff3489974415873426188c71c613d2d28de6e3).
+
+## Building U-Boot
+
+{% include image.html path="/assets/images/content/building-uboot-image-1.png" alt="Building Uboot image 1" %}
+
+## Running QEMU
+
+Make sure swtpm is installed and running on your system. For Debian and friends there's a swtpm package, so just do
+
+{% include image.html path="/assets/images/content/running-qemu-image-1.png" alt="Running EQMU image 1" %}
+
+and launch QEMU with swtpm support
+
+{% include image.html path="/assets/images/content/running-qemu-image-2.png" alt="Running EQMU image 2" %}
+
+## Booting linux
+
+From U-Boot's command line do something along the lines of
+
+{% include image.html path="/assets/images/content/booting-linux-image-1" alt="Booting Linux image 1" %}
```

Comment:
@EbbaSimpson you're missing a file extension here.

Reply:
Sorry what do you mean? I'm trying to add an image.

Reply:
Ignore me I get what you're saying, I have fixed it now.